### PR TITLE
Add Tantivy

### PR DESCRIPTION
**Pull request by ArkinDharawat**
Changes:

- Added script to index TREC documents with Tantivy
- Sparse retriever now uses Tantivy instead of indri.
- Removed files and updated README.md

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -19,36 +19,24 @@ RUN apt-get update && apt-get install -y \
...
-    && tar -xvzf indri-5.11.tar.gz \
-    && rm indri-5.11.tar.gz
-
-# Install Indri.
-RUN cd indri-5.11 \
-    && ./configure CXX="g++ -D_GLIBCXX_USE_CXX11_ABI=0" \
-    && make \
-    && make install
-
 # Install python 3.6 because 3.5 version is not compatible with new certifi changes. Gives error at runtime.
```
Comment (Amit-GH) [inline Dockerfile]: nit: the comment should say python 3.7

File: Dockerfile [hunk 2]
```diff
@@ -19,36 +19,24 @@ RUN apt-get update && apt-get install -y \
...
     && make \
     && make install
 
 # Install pip but not to the latest version as it does not support pyndri installation due to Python 2.7 incompatibility.
 # https://stackoverflow.com/questions/65896334
-RUN apt-get update && apt-get install -y python3-pip \
-    && pip3 install --upgrade "pip < 21.0"
-
-RUN pip3 install pyndri
+#RUN apt-get update && apt-get install -y python3-pip \
```
Comment (Amit-GH) [inline Dockerfile]: remove commented code.

File: README.md [hunk 3]
```diff
@@ -33,120 +36,145 @@ is presented below:
...
 docker exec -it macaw_test_container /bin/bash
 ```
 
 #### Updating TREC data for Indri
-Indri index is created using the document stored in `trec_documents/` directory. It has some default data. To create a bigger
-index, download the entire data from [archive](https://archive.org/details/trec-ir) and put it in trec_documents. Docker will 
-copy it during build time and create a new index. Index creation parameters like *-memory*, *-stemmer.name* can be changed
-inside the Dockerfile.
+
+Indri index is created using the document stored in `trec_documents/` directory. It has some default data. To create a
```
Comment (Amit-GH) [inline README.md]: nit: Tantivy index is created...

File: README.md [hunk 4]
```diff
@@ -33,120 +36,145 @@ is presented below:
...
+bigger index, download the entire data from [archive](https://archive.org/details/trec-ir) and put it in trec_documents.
+Docker will copy it during build time and create a new index. Index creation parameters like *-memory*, *-stemmer.name*
+can be changed inside the Dockerfile.
 
 ### A note on dependencies
-Pyndri is tested only with [Indri-5.11](https://sourceforge.net/projects/lemur/files/lemur/indri-5.11/) which in turn only
-supports certain OS. We use Ubuntu-16.04 in our Docker container. Ubuntu 18.04 creates issues during indri index creation
-because of some unsupported C++11 functions. *pip* version <21.0 is needed to properly install pyndri.
 
+Pyndri is tested only with [Indri-5.11](https://sourceforge.net/projects/lemur/files/lemur/indri-5.11/) which in turn
```
Comment (Amit-GH) [inline README.md]: Pyndri is no longer needed so you can modify this dependency note. Add any new things that are still relevant.

File: macaw/core/retrieval/__init__.py [hunk 5]
```diff
@@ -4,7 +4,8 @@
 Authors: Hamed Zamani (hazamani@microsoft.com)
 """
 import macaw.core.retrieval.bing_api
-import macaw.core.retrieval.indri
+# import macaw.core.retrieval.indri
```
Comment (Amit-GH) [inline macaw/core/retrieval/__init__.py]: nit: remove comment

File: macaw/core/retrieval/__init__.py [hunk 6]
```diff
@@ -28,17 +29,16 @@ def get_retrieval_model(params):
...
 
     params['logger'].info('The search engine for retrieval: ' + params['search_engine'])
-    if params['search_engine'] == 'indri':
-        return macaw.core.retrieval.indri.Indri({'query_generation': q_generation,
-                                    'indri_path': params['search_engine_path'],
-                                    'index': params['col_index'],
-                                    'text_format': params['col_text_format'],
-                                    'results_requested': params['results_requested'],
-                                    'logger': params['logger']})
+    if params['search_engine'] == 'indri' or params['search_engine'] == 'tantivy':
```
Comment (Amit-GH) [inline macaw/core/retrieval/__init__.py]: remove 'indri' from a valid search_engine argument.

File: macaw/core/retrieval/__init__.py [hunk 7]
```diff
@@ -28,17 +29,16 @@ def get_retrieval_model(params):
...
-                                            'bing_key': params['bing_key'],
-                                            'results_requested': params['results_requested'],
-                                            'logger': params['logger']})
+                                                            'bing_key': params['bing_key'],
+                                                            'results_requested': params['results_requested'],
+                                                            'logger': params['logger']})
     else:
-        raise Exception('The requested retrieval model does not exist!')
\ No newline at end of file
+        raise Exception('The requested retrieval model does not exist!')
```
Comment (Amit-GH) [inline macaw/core/retrieval/__init__.py]: nit: You can include the name of invalid search_engine in this error message.

File: macaw/core/retrieval/tantivy.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
+"""
+The Indri search engine.
```
Comment (Amit-GH) [inline macaw/core/retrieval/tantivy.py]: Tantivy

File: macaw/live_main.py [hunk 9]
```diff
@@ -79,9 +79,10 @@ def run(self):
     # These are parameters used by the retrieval model.
     retrieval_params = {'query_generation': 'simple',  # the model that generates a query from a conversation history.
                         'use_coref': True,  # True, if query generator can use coreference resolution, otherwise False.
-                        'search_engine': 'indri',  # the search engine. It can be either 'indri' or 'bing'.
+                        'search_engine': 'tantivy',  # the search engine.
                         'bing_key': 'YOUR_BING_SUBSCRIPTION_KEY',  # Bing API key
-                        'search_engine_path': '/usr/src/app/indri-5.11',  # The path to the indri toolkit.
+                        # 'search_engine_path': '/usr/src/app/indri-5.11',  # The path to the indri toolkit.
```
Comment (Amit-GH) [inline macaw/live_main.py]: can simply delete this.

File: macaw/core/retrieval/tantivy.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            docs = [get_trec_doc(self.searcher.doc(doc_id)['body'][0])
+                    for _, doc_id in scores]  # convert the raw text into trec-doc
+        except Exception as e:
+            self.logger.error(f'Error on Query {e}')
+            return None
+
+        return docs
+
+    def get_doc_from_index(self, doc_id):
+        # TODO: document should be TREC doc here
```
Comment (Amit-GH) [inline macaw/core/retrieval/tantivy.py]: What does this todo mean? Do you want to verify that doc_id belongs to a valid trec document?

**Review by Amit-GH — CHANGES_REQUESTED** [review]
Changes look good. Most comments should be easily resolved. In the Readme file, we can aim to keep the Docker section updated (it shouldn't have any reference to indri). The manual installation part can be outdated for now. To fix that, someone should try that installation.

File: macaw/core/retrieval/tantivy.py (hunk 10 above)
Comment (ArkinDharawat) [reply macaw/core/retrieval/tantivy.py]: Old TODO, that change is already implemented

**Comment by ArkinDharawat** [comment]
Made all the requested changes.

**Review by Amit-GH — APPROVED** [review]

**Pull request merged**